### PR TITLE
PTX-4504 Add delete pods by labels and deployment functions

### PR DESCRIPTION
**Pull request by nikolaypopov**
Add delete pods by labels and deployment functions.

### CODE REVIEW

File: k8s/core/pods.go [hunk 1]
```diff
@@ -297,6 +273,59 @@ func (c *Client) GetPodByUID(uid types.UID, namespace string) (*corev1.Pod, erro
...
+		return err
+	}
+
+	deleteOptions := metav1.DeleteOptions{}
+	if force {
+		gracePeriodSec := int64(0)
+		deleteOptions.GracePeriodSeconds = &gracePeriodSec
+	}
+
+	return c.kubernetes.CoreV1().Pods(ns).Delete(context.TODO(), name, deleteOptions)
```
Comment (piyush-nimbalkar) [inline k8s/core/pods.go:299]: Let's call the methods from `common` here so we don't duplicate code.

File: k8s/apps/deployments.go [hunk 2]
```diff
@@ -293,3 +296,35 @@ func (c *Client) GetDeploymentsUsingStorageClass(scName string) ([]appsv1.Deploy
...
+	logrus.Debugf("Deleting pods: %s", podsNamesToDelete)
+	if err := common.DeletePods(c.core, pods, false); err != nil {
+		return err
+	}
+
+	if err := common.WaitForPodsToBeDeleted(c.core, podsToDelete, timeout); err != nil {
+		return fmt.Errorf("Failed to wait for pods to be deleted: %s, Err: %v", podsNamesToDelete, err)
+	}
+
+	logrus.Debugf("Pods for deployment %s have been successfully deleted", name)
```
Comment (piyush-nimbalkar) [inline k8s/apps/deployments.go:328]: Let's not log from here for consistency. Also, we don't know if every caller is using logrus or not. The caller can always print these two logs before and after their call.

File: k8s/core/pods.go [hunk 3]
```diff
@@ -297,6 +273,59 @@ func (c *Client) GetPodByUID(uid types.UID, namespace string) (*corev1.Pod, erro
...
+	logrus.Debugf("Deleting pods with labels %v: %s", listOptions, podsNamesToDelete)
+	if err := c.DeletePods(pods.Items, false); err != nil {
+		return err
+	}
+
+	if err := common.WaitForPodsToBeDeleted(c.kubernetes.CoreV1(), podsToDelete, timeout); err != nil {
+		return fmt.Errorf("Failed to wait for pods to be deleted: %s, Err: %v", podsNamesToDelete, err)
+	}
+
+	logrus.Debugf("Pods with labels %v have been successfully deleted", listOptions)
```
Comment (piyush-nimbalkar) [inline k8s/core/pods.go:325]: Same comment about logging as above.

File: k8s/apps/deployments.go (hunk 2 above)
Comment (nikolaypopov) [reply k8s/apps/deployments.go]: Done, remove logrus.Debugf()

File: k8s/core/pods.go (hunk 3 above)
Comment (nikolaypopov) [reply k8s/core/pods.go]: Done, remove logrus.Debugf()

File: k8s/core/pods.go (hunk 1 above)
Comment (nikolaypopov) [reply k8s/core/pods.go]: Calling GetPods() from the common now, GetPod() is not in common

File: k8s/common/pod.go [hunk 4]
```diff
@@ -48,6 +53,118 @@ func GeneratePodsOverviewString(pods []corev1.Pod) string {
...
+	if len(in) == 0 {
+		return ""
+	}
+	return labels.FormatLabels(in)
+}
+
+func getPodsWithListOptions(client v1.CoreV1Interface, namespace string, opts metav1.ListOptions) (*corev1.PodList, error) {
+	return client.Pods(namespace).List(context.TODO(), opts)
+}
+
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply k8s/common/pod.go]: Done

File: k8s/common/pod.go [hunk 5]
```diff
@@ -48,6 +53,118 @@ func GeneratePodsOverviewString(pods []corev1.Pod) string {
...
+	if !ok {
+		return fmt.Errorf("Failed to delete pods")
+	}
+	return nil
+}
+
+// WaitForPodDeletion waits for given timeout for given pod to be deleted
+func WaitForPodDeletion(client v1.CoreV1Interface, uid types.UID, namespace string, timeout time.Duration) error {
+	t := func() (interface{}, bool, error) {
+		p, err := GetPodByUID(client, uid, namespace)
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply k8s/common/pod.go]: Done

File: k8s/common/pod.go [hunk 6]
```diff
@@ -48,6 +53,118 @@ func GeneratePodsOverviewString(pods []corev1.Pod) string {
...
+	}
+
+	go func() {
+		wg.Wait()
+		close(errChan)
+	}()
+
+	ok := true
+	for err := range errChan {
+		logrus.Error(err)
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply k8s/common/pod.go]: Done

File: k8s/common/pod.go [hunk 7]
```diff
@@ -48,6 +53,118 @@ func GeneratePodsOverviewString(pods []corev1.Pod) string {
...
+		go func(pod corev1.Pod, index int) {
+			defer wg.Done()
+			if err := WaitForPodDeletion(client, pod.UID, pod.Namespace, timeout); err != nil {
+				errChan <- fmt.Errorf("Failed to delete pod %s, Err: %v", pod.Name, err)
+			}
+		}(pod, index)
+
+	}
+
+	go func() {
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply k8s/common/pod.go]: Done

File: k8s/apps/deployments.go [hunk 8]
```diff
@@ -36,6 +37,8 @@ type DeploymentOps interface {
 	DescribeDeployment(name, namespace string) (*appsv1.DeploymentStatus, error)
 	// GetDeploymentsUsingStorageClass returns all deployments using the given storage class
 	GetDeploymentsUsingStorageClass(scName string) ([]appsv1.Deployment, error)
+	// DeletePodsByDeployment deletes pods for the given deployment name and namespace
+	DeletePodsByDeployment(name, namespace string, timeout time.Duration) error
```
[the remark this answers is not in the document]
Comment (nikolaypopov) [reply k8s/apps/deployments.go]: Renamed

**Review by piyush-nimbalkar — APPROVED** [review]

**Pull request merged**